tests/FSharp.Compiler.ComponentTests/EmittedIL/ComputedCollections/ForNInRangeArrays.fs.il.bsl

```diff
@@ -0,0 +1,2493 @@
+    IL_0013:  conv.i
+    IL_0014:  ldloc.2
+    IL_0015:  stloc.3
+    IL_0016:  ldarg.0
+    IL_0017:  ldnull
+    IL_0018:  callvirt   instance !1 class [FSharp.Core]Microsoft.FSharp.Core.FSharpFunc`2<class [FSharp.Core]Microsoft.FSharp.Core.Unit,class [FSharp.Core]Microsoft.FSharp.Core.Unit>::Invoke(!0)
+    IL_001d:  pop
+    IL_001e:  ldloc.3
+    IL_001f:  stelem.i4
+    IL_0020:  ldloc.2
+    IL_0021:  ldc.i4.1
+    IL_0022:  add
+    IL_0023:  stloc.2
+    IL_0024:  ldloc.1
+    IL_0025:  ldc.i4.1
+    IL_0026:  conv.i8
+    IL_0027:  add
+    IL_0028:  stloc.1
+    IL_0029:  ldloc.1
+    IL_002a:  ldc.i4.s   10
+    IL_002c:  conv.i8
+    IL_002d:  blt.un.s   IL_0011
+
+    IL_002f:  ldloc.0
+    IL_0030:  ret
+  } 
+
+  .method public static int32[]  f00(class [FSharp.Core]Microsoft.FSharp.Core.FSharpFunc`2<class [FSharp.Core]Microsoft.FSharp.Core.Unit,class [FSharp.Core]Microsoft.FSharp.Core.Unit> f,
+                                     class [FSharp.Core]Microsoft.FSharp.Core.FSharpFunc`2<class [FSharp.Core]Microsoft.FSharp.Core.Unit,class [FSharp.Core]Microsoft.FSharp.Core.Unit> g) cil managed
+  {
+    .custom instance void [FSharp.Core]Microsoft.FSharp.Core.CompilationArgumentCountsAttribute::.ctor(int32[]) = ( 01 00 02 00 00 00 01 00 00 00 01 00 00 00 00 00 ) 
+    
+    .maxstack  6
+    .locals init (int32[] V_0,
+             uint64 V_1,
+             int32 V_2,
+             int32 V_3)
+    IL_0000:  ldc.i4.s   10
+    IL_0002:  conv.i8
+    IL_0003:  conv.ovf.i.un
+    IL_0004:  newarr     [runtime]System.Int32
+    IL_0009:  stloc.0
+    IL_000a:  ldc.i4.0
+    IL_000b:  conv.i8
+    IL_000c:  stloc.1
+    IL_000d:  ldc.i4.1
+    IL_000e:  stloc.2
+    IL_000f:  br.s       IL_0031
+
+    IL_0011:  ldloc.0
+    IL_0012:  ldloc.1
+    IL_0013:  conv.i
+    IL_0014:  ldloc.2
+    IL_0015:  stloc.3
+    IL_0016:  ldarg.0
+    IL_0017:  ldnull
+    IL_0018:  callvirt   instance !1 class [FSharp.Core]Microsoft.FSharp.Core.FSharpFunc`2<class [FSharp.Core]Microsoft.FSharp.Core.Unit,class [FSharp.Core]Microsoft.FSharp.Core.Unit>::Invoke(!0)
+    IL_001d:  pop
+    IL_001e:  ldarg.1
+    IL_001f:  ldnull
+    IL_0020:  callvirt   instance !1 class [FSharp.Core]Microsoft.FSharp.Core.FSharpFunc`2<class [FSharp.Core]Microsoft.FSharp.Core.Unit,class [FSharp.Core]Microsoft.FSharp.Core.Unit>::Invoke(!0)
+    IL_0025:  pop
+    IL_0026:  ldloc.3
+    IL_0027:  stelem.i4
+    IL_0028:  ldloc.2
+    IL_0029:  ldc.i4.1
+    IL_002a:  add
+    IL_002b:  stloc.2
+    IL_002c:  ldloc.1
+    IL_002d:  ldc.i4.1
+    IL_002e:  conv.i8
+    IL_002f:  add
+    IL_0030:  stloc.1
+    IL_0031:  ldloc.1
+    IL_0032:  ldc.i4.s   10
+    IL_0034:  conv.i8
+    IL_0035:  blt.un.s   IL_0011
+
+    IL_0037:  ldloc.0
+    IL_0038:  ret
+  } 
+
+  .method public static int32[]  f000(class [FSharp.Core]Microsoft.FSharp.Core.FSharpFunc`2<class [FSharp.Core]Microsoft.FSharp.Core.Unit,class [FSharp.Core]Microsoft.FSharp.Core.Unit> f) cil managed
```

Comment:
Just curious: This one keeps using RangeInt32 + allocating ListCollector because of having 2 yields inside?

Reply:
Yep.

> We could theoretically go even further and detect a statically-known number of `yield`s > 1 and factor that back into the total iteration count, but that seems like a step too far for now.

Reply:
Understand.

It might make sense eventually if code like this with fixed number of unbranched yields is the case in some popular GUI-building library for example.

Reply:
> It might make sense eventually if code like this with fixed number of unbranched yields is the case in some popular GUI-building library for example.

...Although, without further cleverness, factoring additional `yield`s into the `count` could technically be a runtime breaking change if someone were doing something extremely odd like this —

```fsharp
let f n =
    if n >= 100u then raise (BreakOutOfTheLoopException "Breaking out of the loop.")
    else n

try
    [| for n in UInt32.MinValue .. UInt32.MaxValue / 3u do
        yield f n
        yield f n |]
with
// This would be hit before.
| :? BreakOutOfTheLoopException as e -> …
// Now this would be hit when we
// tried to initialize the array with `count`
// after factoring in the additional yields,
// since (2³² / 3 + 1) × 2 > System.Array.MaxLength.
| e -> …
```

— since, with the `count` trick as-is, an overflow exception would be raised when trying to initialize the array with $count = (\frac{2^{32}}{3} + 1) \times 2 > System.Array.MaxLength$.

Something similar would apply for $count > 2^{64} + 1$, e.g.,

```fsharp
let f n = printfn $"{n}"; n

[ for n in UInt64.MinValue..UInt64.MaxValue do
    yield f n
    yield f n ]
```

which, before, would print `n` until it ran out of memory, whereas now `count` would be $(2^{64} + 1) \times 2$, which would raise an overflow exception before `n` was ever printed.

But I can't really think of a legitimate reason why anyone would depend on code like that... right??

Reply:
It's only in a degenerate scenario and it is behind a feature flag already, this would be fine.
And agree, let's get this PR in first.